add e2e test for custom token send transfer action

mirjanaKukic opened the pull request:
## Explanation
We need to refactor these tests since they already exist in _metamask-ui.spec.js_ file and we want them to use fixtures.
New tests will be moved to _custom-token-send-transfer.spec.js_ file. 
Once all refactoring is done these tests will be removed from metamask-ui.spec.js

Resolves [#13676](https://github.com/MetaMask/metamask-extension/issues/13676)

## More Information

The main goal is to deploy the e2e smart contracts directly on Ganache, instead of using the Dapp Test interface for achieving the same actions.
These smart contracts can be used later within the e2e tests in order to reduce the amount of code needed to be written in order to execute the same action every time for each of the tests (i.e. contract deployment). This would also reduce the time needed for e2e tests execution because waiting for the UI action response will be eliminated.

## Manual Testing Steps

Run testcase locally or check ci pipeline and ensure that it continues to work normally:
`yarn test:e2e:single test/e2e/tests/custom-token-send-transfer.spec.js`

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [ ] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

mirjanaKukic committed:
add send token inside metamask

send token inside metamask

fix

changed wait

send token option

fix

added delays

delay

changed delay

transfer token without specifying gas

fix

fix

fix

lint fix

assert changed

added waitforselector

changed delay

add waitforselector

add delay

added waitForSelector

lint:fix

fix

added failOnConsoleError

changed delays

fix

removed delays

changed waitforselector

fix

fix

lint fix

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,414 @@
+
+        await driver.switchToWindow(windowHandles.extension);
+
+        const assets = await driver.waitForSelector({
```
Should this be `asset`? since its only one

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,414 @@
+        );
+        await driver.clickElement('[data-testid="page-container-footer-next"]');
+
+        await driver.delay(2000);
```
Why do we need this delay? Can't we use waitForSelector?

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,414 @@
+        windowHandles = await getWindowHandles(driver, 3);
+        await driver.switchToWindow(windowHandles.popup);
+
+        await driver.delay(4000);
```
Why do we need this delay?

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,414 @@
+        windowHandles = await getWindowHandles(driver, 3);
+        await driver.switchToWindow(windowHandles.popup);
+
+        await driver.delay(4000);
```
Why do we need this delay?

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,414 @@
+        await driver.clickElement({
+          text: 'Activity',
+          tag: 'button',
+        });
```
Any reason why we are ending the test with an action instead of an assertion? Why do we need to click on this element?

PeterYinusa reviewed (dismissed):
Can you use the [ganache seeder](https://github.com/MetaMask/metamask-extension/blob/develop/test/e2e/seeder/ganache-seeder.js) to deploy the hst contract, so you do not have to deploy the contract using the test-dapp in each test. This will also help remove a lot of code from each test

mirjanaKukic committed:
refactoring to use ganache seeder

added comments

fix

removed delay

small refactor

fix

lint fix

code refactor

changed description of describe blocks

fix

fix

added fixture builder in test

seaona commented:
Same comment as [here](https://github.com/MetaMask/metamask-extension/pull/15728#issuecomment-1238141179). What do you think @PeterYinusa @mirjanaKukic

seaona commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,445 @@
+          'MetaMask Notification',
+          windowHandles,
+        );
+        await driver.waitForSelector({ text: '1.5 TST', tag: 'h1' });
```
I think we can remove this line

mirjanaKukic replied:
If I remove it I'm getting error on the line after it : 
 _ElementClickInterceptedError: Element <button class="button btn--rounded btn-primary page-container__footer-button"> is not clickable at point (266,567) because another element <div class="loading-overlay"> obscures it_

So I put this waitForSelector here since it seems like it needs a little bit more time to load popup properly. I could use delay but we all agreed to avoid it. Hope this makes sense.

seaona replied:
Thank you, that make sense! Let's leave it as it is then! @mirjanaKukic

seaona commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,445 @@
+    await withFixtures(
+      {
+        dapp: true,
+        fixtures: 'connected-state',
```
there is a fixture for a `custom-token`, could we use it instead of this one? This might help reduce the test steps, as we won't need to add the token to the wallet then

mirjanaKukic replied:
Good idea. I tried it but when I run test it shows me 0 TST so I'm not able to send tokens since I don't have enough tokens. Maybe there is way to change custom-token fixture in order to have more TST tokens?

seaona commented:
Maybe we could summarize `it` blocks and `describes`? At the moment this is the current text and hierarchy:
```
describe('Send token from inside MetaMask')
    it('starts to send a transaction, transitions to the confirm screen, displays the token transfer data, customizes gas, submits the transaction, finds the transaction in the transactions list'

describe('Send a custom token from dapp')
    it('sends an already created token')
    it('customizes gas, submits the transaction and finds the transaction in the transactions list')
    
describe('Transfers a custom token from dapp when no gas value is specified')
    it('transfers an already created token without specifying gas, submits the transaction, finds the transaction in the transactions list')
```
A proposal for simplifying tests (we could reduce code by removing common values on the different `describes` by having just one) and improve test-redability a little bit. 100% open for discussion @mirjanaKukic @PeterYinusa  

```
describe('Custom Token Transactions')
    it('can be sent from inside MetaMask')
    it('can be transferred from a dapp')
    it('can be transferred from a dapp customizing gas values')
    it('can be transferred without specifying gas from a dapp'')
```

seaona commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,445 @@
+          tag: 'h1',
+          text: '1.5 TST',
+        });
+        await driver.waitForSelector({ text: 'Confirm', tag: 'button' });
```
I think we can remove this line

mirjanaKukic replied:
Same as above. If I remove it test fails since it needs more time to load popup.

seaona replied:
Thank you, that make sense! Let's leave it as it is then! @mirjanaKukic

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,420 @@
+        });
+        asset.click();
+        // send token from extension
+        await driver.waitForSelector('[data-testid="eth-overview-send"]');
```
You shouldn't need to use waitForSelector followed by clickElement, as clickElement waits until the element is located

PeterYinusa replied:
Thanks for the explanation, this issue occurs because we use the same selector on both the overview page and the asset overview page. Instead of `await driver.waitForSelector('[data-testid="eth-overview-send"]');` can we wait for the overview page to transition to the asset overview page e.g. wait for the balance to show 10 TST?

mirjanaKukic replied:
Hmm, I removed it and it works now. Tested it locally and also works after removing waitForSelector.

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,420 @@
+        // continue to next screen
+        await driver.waitForSelector(
+          '[data-testid="page-container-footer-next"]',
+        );
```
same as above

mirjanaKukic replied:
Done

brad-decker commented:
@PeterYinusa @seaona this is ready for another look.

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,426 @@
+          css: '.list-item',
+          text: '10 TST',
+        });
+        asset.click();
```
Instead of using `waitForSelector` followed by `click`, can we just use `clickElement` as the asset variable is only used once

mirjanaKukic replied:
Done

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,426 @@
+        await driver.clickElement('[data-testid="page-container-footer-next"]');
+        // added wait for selector since it needs some time to load this page
+        // otherwise would have to use delays
+        await driver.waitForSelector({ text: 'Edit', tag: 'button' });
```
we should be able to remove `await driver.waitForSelector({ text: 'Edit', tag: 'button' });`

mirjanaKukic replied:
Done

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,426 @@
+        await driver.waitForSelector({
+          css: '.list-item__heading',
+          text: 'Send TST',
+        });
```
We should be able to remove the code block below, as theres already a wait for the completed transaction
```
await driver.waitForSelector({
          css: '.list-item__heading',
          text: 'Send TST',
        });
```

mirjanaKukic replied:
Done

PeterYinusa commented on test/e2e/tests/custom-token-send-transfer.spec.js:
```diff
@@ -0,0 +1,426 @@
+        await driver.waitForSelector({
+          css: '.transaction-list-item__primary-currency',
+          text: '-1.5 TST',
+        });
```
We should be able to remove the following code block, as the line that follows waits for the completed transaction list item
```
        await driver.wait(async () => {
          const pendingTxes = await driver.findElements(
            '.transaction-list__pending-transactions .transaction-list-item',
          );
          return pendingTxes.length === 1;
        }, 10000);

        await driver.waitForSelector({
          css: '.transaction-list-item__primary-currency',
          text: '-1.5 TST',
        });
```

mirjanaKukic replied:
Done

mirjanaKukic committed:
fix

mirjanaKukic committed:
removed waitForSelector

chloeYue reviewed (approved):
LGTM !

brad-decker commented:
@PeterYinusa do we want this PR still and if so who should own it?

PeterYinusa commented:
> @PeterYinusa do we want this PR still and if so who should own it?

Yes, we should get this in. I'll review it shortly and get it merged. We can do a follow-up PR to clean up anything outstanding if needed.
